Feature/join 2018 (#451)
ordina-jworks/ordina-jworks.github.io: _posts/2018-11-16-JOIN-2018.md

#### Tic Tac Toe
In contrary to [Go](https://en.wikipedia.org/wiki/Go_(game)){:target="_blank"}, Tic Tac Toe just has 360.000 possible moves making it easier to use for a demo.

The network for this demo learned by playing against some random opponents without any real strategy, this made it quite dumb. 
You could improve this by letting it play against itself.

#### Digital Railway Survey
The idea is to recognize signs next to the railway and verify if these have been installed at the correct place.

[You Only Look Once](https://en.wikipedia.org/wiki/Object_detection){:target="_blank"} is a very cool algorithm which gives you very nice bounded boxes around the objects on the images making it easier for your network to know where it needs to look and thus preventing you from processing unrelevant pictures.

It was also nice to see that with deep learning the system can distinguish signs from a passing train even when we would be unable to see these signs, let alone see what has been written on it.

### Questions
#### How long does it take to get started?
If you work with existing data sets and known algorithms, 15 lines of code is enough.

> For a brand new project it is easy up to 55 days just to show that deep learning is possible for a specific business case.

Author: ```suggestion
Remco mentioned that for a brand new project it can easily take up 55 days just to show that deep learning is possible for a specific business case.
```

> Remco mentioned that for a brand new project it can easily take up 55 days just to show that deep learning is possible for a specific business case.

Author: ```suggestion
Remco mentioned that for a brand new project it can easily take up to 55 days just to show that deep learning is possible for a specific business case.
```

## Unbeatable at Connect 4 with Wearables and Computer Vision by [Frederick Bousson](https://twitter.com/fbousson){:target="_blank"}

<span class="image left"><img class="p-image" alt="Frederick Bousson" src="/img/join-2018/speaker-bousson.jpg"></span>

Frederick Bousson kicked off the technology track with his talk about wearables.

Wearables are becoming more mainstream every day as proven by the prevalence of fitness trackers and smart watches.
Smart glasses on the other hand are still very rare.
Mainly because they still look awful.
But improvement is on its way here as well: better looking sets are starting to appear like the [Vuzix Blade](https://www.vuzix.com/products/blade-smart-glasses){:target="_blank"}.
This is important as humans are very vision-oriented: 90% of the information on the internet is processed via the eyes.

### Connect 4
To really demonstrate the power of smart glasses, Frederick was looking for an interesting case he could use as a demonstration.
The result of this quest was a way to help you become unbeatable at Connect 4.

Connect 4 is a "solved game", which means that there's an algorithm that can't be defeated (provided you get to go first).

> Frederick went hunting in the Open Source community and discovered a program written by [Brian Borowski](http://www.brian-borowski.com/) which contained this algorithm. 

Reviewer: ```suggestion
Frederick went hunting in the Open Source community and discovered a program written by [Brian Borowski](http://www.brian-borowski.com/){:target="_blank"} which contained this algorithm. 
```

4 years ago, [Scott Bouloutian](https://github.com/ScottBouloutian/OpenCV-Connect-Four-AI){:target="_blank"} added visual processing code on top of that and yet another person ported the result to Google Glass.
Nice to see how, with open source, you have people who build upon each others' work.

### Image processing
If you want to have your smart glasses solve your game of Connect 4, you'll need to do some image processing.

> And since the game revolves around colour which means your program doesn't just need to see, it has to reason about these colors as well.

Reviewer: ```suggestion
And since the game revolves around colour which means your program doesn't just need to see, it has to reason about these colours as well.
```

If your brain is conditioned to think in RGB, switching to BGR is not an easy task.

Before you can start solving the game you need to know the position of the pieces that were already played.
And before you can find those pieces you need to know where the board is.
In order to find the board you'll need to look for a large, blue rectangle.
Then you draw lines around it and find the corners.
Now your application "sees" the board.
<div style="text-align: center;">
  <img src="/img/join-2018/draw-boxes.jpg" width="100%">
</div>

The next step is to look at the location of the played pieces.
For this you need to look inside the board for concentrations of the right color.
If you're using a board that has a tray for the game pieces this tray is also part of the board and you'll need to cut off a bit from the bottom.

Of course if you're using a different light source, colours are perceived differently by the camera (e.g. the same colour in fluorescent lighting is green while it will be matched to yellow when using a light bulb). 

> This means that the colors found by openCV might no longer match the ones you programmed in so you'll also have to take the white balance into account.

Reviewer: ```suggestion
This means that the colours found by openCV might no longer match the ones you programmed in so you'll also have to take the white balance into account.
```

TVH is a company which rents out equipment and can provide you with spare parts for a lot of tools and vehicles.

### MyAssetPlanner

[Tomorrowland](https://www.tomorrowland.com/){:target="_blank"} is quite a big event and for setting up and tearing down they rent up to 160 machines (or "assets").
For every asset a contractor will need to schedule it to be available at a certain time, resulting in more than 300 phone calls which need to be handled by the asset planner.

TVH wanted to improve the customer experience and decided they were going to develop a new application to help with the planning of these assets: MyAssetPlanner was born.

First there was a design sprint by [Clockwork](https://clockwork.ordina.be/){:target="_blank"} in order to come up with a prototype and the most important business needs which needed to be implemented. 
This part took about two weeks.

After this prototype, TVH started with the implementation of the first version which was put into production two months later.
This first version has currently been in use for four months including for the last iteration of Tomorrowland.

> For development the [Spotify model](https://labs.spotify.com/2014/03/27/spotify-engineering-culture-part-1/){:target="_blank"} was used which is focussed around:

Reviewer: ```suggestion
For development, the [Spotify model](https://labs.spotify.com/2014/03/27/spotify-engineering-culture-part-1/){:target="_blank"} was used which is focussed around:
```

Very soon this will scale up to 5% of all TVH customers which amounts to 3.000.
After that there will be a gradual ramp up towards 20.000 customers.

> You can rewatch his [talk](https://www.youtube.com/watch?v=K396nSVtfrE){:target="_blank"} on our [YouTube channel](https://www.youtube.com/channel/UCsebfWdqV7LqNNDMDvCESIA){:target="_blank"}.

## Visualizing Mandelbrot fractals using Riff and Spring Cloud Function by [Dieter Hubau](https://twitter.com/dhubau){:target="_blank"}

<span class="image left"><img class="p-image" alt="Dieter Hubau" src="/img/join-2018/speaker-hubau.jpg"></span>

#### Function as a Service
[Pivotal Function Service](https://pivotal.io/platform/pivotal-function-service){:target="_blank"} (PFS) is a pretty new addition to the Pivotal landscape.

[Riff](https://projectriff.io/){:target="_blank"}, an open source project, is the base of PFS.
The idea behind Riff is to easily create functions and move these to the cloud.
Its name is derived from the guitar world as you can see [here](https://en.wikipedia.org/wiki/Ostinato#Riff){:target="_blank"}, the project lead is [Mark Fisher](https://www.twitter.com/m_f_){:target="_blank"}.

> Initially Zookeeper and Kafka were used to send messages between functions, but this did mean a lot of extra overhead to set up.

Reviewer: ```suggestion
Initially Zookeeper and Kafka were used to send messages between functions but this did mean a lot of extra overhead to set up.
```

### Mandelbrot
One of the simplest mathematical functions which creates the most complex objects is the [Mandelbrot](https://en.wikipedia.org/wiki/Mandelbrot_set){:target="_blank"} set.

You can keep zooming in, resulting in more calculations the deeper you go.
It also tends to create pretty graphics which is nice for a demo.

### The code
For a 400 x 400 image we will send a request to the backend to calculate the result for every pixel.

This means that around 160.000 calls will be sent to do the calculation.
It is not the most efficient way to do the calculation but this demo is about using functions and not about calculating a Mandelbrot really fast.

Riff uses containers to package your business logic.
There are two kinds:
* UserContainer: actual container contains the actual function
* InitContainer: they run before your container.
You can use these to instantiate a database.

For one function you need about 200MB of memory as it spins up a slimmed down Spring Boot in the backend.
All in all it is still pretty performant.

> Dieter mentioned a talk of Dave Syer about performance in Spring which is very interesting to watch: [link](https://www.youtube.com/watch?v=97UTDmonq7w){:target="_blank"}

Reviewer: ```suggestion
Dieter mentioned a talk of Dave Syer about performance in Spring which is very interesting to watch: [link](https://www.youtube.com/watch?v=97UTDmonq7w){:target="_blank"}.
```

## Testimonial: HZIV by [Pieter Van Hees](https://twitter.com/pvanhees91){:target="_blank"}

<span class="image left"><img class="p-image" alt="Pieter Van Hees" src="/img/join-2018/speaker-pieter.jpg"></span>

Pieter is currently on a project at [HZIV](https://www.caami-hziv.fgov.be/en){:target="_blank"}, a government health agency which offers its members legal health and disability insurance.

### The old situation
A lot of old applications still run with the almighty [Cobol](https://en.wikipedia.org/wiki/COBOL){:target="_blank"}.

10 years ago, HZIV wanted to rewrite these using Java 5 and Swing but that rewrite didn't go so well.
So they are still using these old beasts.

### A new start
With more time comes more insight and it was decided to do a new rewrite and also to give the teams more freedom in how to set up their development environment and choosing the technologies that are best suited.

Chosen technologies are, most notably:
* Java 8 & 10
* [MongoDB](https://www.mongodb.com/){:target="_blank"}
* [ELK](https://www.elastic.co/elk-stack){:target="_blank"}

> Currently the teams are working in an agile way delivering fancy new features.

Reviewer: ```suggestion
Currently the teams are working in an agile way, delivering fancy new features.
```

### Working @ HZIV
#### What is great?
You have great freedom to choose and test out new stuff. 
At the same time the atmosphere is relaxed, allowing you to explore without pressure.

#### What is bad?
There is the risk of Developer Anarchy, where developers just introduce fancy new stuff without properly ironing out the edges.

> Some managers like to micro-manage 

Reviewer: ```suggestion
Some managers like to micro-manage.
```

> The same managers also tend to introduce frequent scope changes, this does not combine well.

Reviewer: ```suggestion
The same managers also tend to introduce frequent scope changes which does not combine well.
```

#### Wishful thinking
Don't fall in the trap of wishful thinking.
Do not just assume that your product will be a success.

Know how to validate your assumptions and make sure that you have a good feedback loop so that you know that the features you provided fulfil the business requirements.

#### Ignoring the facts
Ignoring the obvious signs that you are not going to meet your goal.

So many demos are made with Powerpoint, a real sprint review should be the user starting the application and using it themselves.
If they get stuck you need to ask them why, as these situations will give you great feedback.

### 7. Obsessing over details
The product owner who used to be an analyst.

A huge backlog is not good, a good exercise is to take the top 50 and delete the rest.

In a huge backlog there is a lot 'relevant' information in the stories, however this information tends to become stale.
Later on it will also become increasingly difficult to implement these stories as there is no certainty that the information they contain is still correct.

Another big no-no is having a 'definition of done' being more than one page long for a user story.

A nice concept is the three C's: 

> * Card (post it - fat pen): a story is just a single card. This is a remindeder that we need to talk about this

Reviewer: ```suggestion
* Card (post it - fat pen): a story is just a single card. This is a reminder that we need to talk about this.
```

* Criteria: Only Acceptance Criteria on the story.
* Conversation: During refinement there will be a conversation with the team and then you will expand the story.

Also be careful not to be a perfectionist.
No first release is going to be perfect.
Being fast in the market beats having a complete product.

### 8. No pressence, no communication
Communication is one of the most important factors of failure in an organization.

> 

Reviewer: Would delete this line.
```suggestion
```

Deep learning has two flows:
* Forward propagation: feed data through it.
* Backward propagation: update weights so the next iteration will be better.

You can define weights on input in order to make certain inputs more important.

Deep learning requires a very large network with a large amount of nodes and a lot of hidden layers.

The early layers allow your network to distinguish basic structures within your data.
Further down in the layers the network will learn more detailed features like; eyes, ears, ... 

It is cool that the network learns these features about how to recoginize a person just by providing it with pictures of faces. 
We did not provide it with the concept `eyes` but eventually, in one of the iterations the network started to take eyes into consideration.

### Deep Learning at Ordina

#### Workshop: hack a drone.
A small drone was equiped with deep learning to detect and identify objects by using the [Deep Learning for Java](https://deeplearning4j.org/){:target="_blank"}.
In a short demo we saw it could identify a notebook.

#### Tic Tac Toe

> In contrary to [Go](https://en.wikipedia.org/wiki/AlphaGo){:target="_blank"}, Tic Tac Toe just has 360.000 possible moves making it easier to use for a demo.

Reviewer: This link still points to the AI, shouldn't it point to the Go bord game instead?
```suggestion
In contrary to [Go](https://en.wikipedia.org/wiki/Go_(game)){:target="_blank"}, Tic Tac Toe just has 360.000 possible moves making it easier to use for a demo.
```

The network for this demo learned by playing against some random opponents without any real strategy, this made it quite dumb. 
You could improve this by letting it play against itself.

#### Digital Railway Survey
The idea is to recognize signs next to the railway and verify if these have been installed at the correct place.

[You Only Look Once](https://en.wikipedia.org/wiki/Object_detection){:target="_blank"} is a very cool algorithm which gives you very nice bounded boxes around the objects on the images making it easier for your network to know where it needs to look and thus preventing you from processing unrelevant pictures.

It was also nice to see that with deep learning the system can distinguish signs from a passing train even when we would be unable to see these signs, let alone see what has been written on it.

### Questions
#### How long does it take to get started?
If you work with existing data sets and known algorithms, 15 lines of code is enough.

> For a brand new project it is easy up to 55 days just to show that deep learning is possible for a specific business case.

Reviewer: I still don't understand the 55 days?

It takes a long time to gather the required data.
You need to label it and make sure it is diverse enough so your model is not too focussed on one specific subset of input.

> You can rewatch his [talk](https://www.youtube.com/watch?v=NcT7Hif19hk){:target="_blank"} on our [channel](https://www.youtube.com/channel/UCsebfWdqV7LqNNDMDvCESIA){:target="_blank"}.

## Innovation without asking permission by [Bart Blommaerts](https://twitter.com/DaggieBe){:target="_blank"}
<span class="image left"><img class="p-image" alt="Bart Blommaerts" src="/img/join-2018/speaker-bart.jpg"></span>

### Service decomposition
When you want to split up a monolith into smaller applications there are three patterns you can follow.

* Split: define vertical functional boundaries.
This does not happen to be possible in all old applications.
* Strangle: extract and re-implement logic in new components.

> * Extend: new functionality in new microservices. Do pay attention to not create a distributed micro-**** fest, also known as a distributed monolith.

Reviewer: Seems like we need four more... :)
```suggestion
* Extend: new functionality in new microservices. Do pay attention to not create a distributed micro-******** fest, also known as a distributed monolith.
```
